src/game-of-life/implementing.md

```diff
@@ -390,6 +390,109 @@ Make sure your development server is still running (run `npm run start` inside
 
 [![Screenshot of the Game of Life implementation with text rendering](../images/game-of-life/initial-game-of-life-pre.png)](../images/game-of-life/initial-game-of-life-pre.png)
 
+## Testing our Game of Life
+
+Now that we have our Rust implementation of the Game of Life rendering in the browser with JavaScript, let's talk about testing our Rust compiled to WebAssembly functions.
+
+We are going to test our `tick` function to make sure that it gives us the output that we expect.
+First, we'll want to create some setter functions inside our `impl Universe` in the `wasm_game_of_life/src/lib.rs` file. We are going to create a `set_width` and a `set_height` function so we can create `Universe`s of different sizes.
+
+```rust
+#[wasm_bindgen]
+impl Universe { 
+    // ...
+
+    pub fn set_width(mut self, width: u32) -> Self {
+        self.width = width;
+        // Reinitialize cells with new width and the universe's height.
+        // Set all the cells to dead.
+        self.cells = (0..width * self.height).map(|_i| Cell::Dead).collect();
+        self
+    }
+
+    pub fn set_height(mut self, height: u32) -> Self {
+        self.height = height;
+        // Reinitialize cells with new height and the universe's width.
+        // Set all the cells to dead.
+        self.cells = (0..self.width * height).map(|_i| Cell::Dead).collect();
+        self
+    }
+
+}
+```
+
+Now we are going to make some public Rust functions for our `Universe`. We aren't going to compile them to WebAssembly, however. Rust compiled into WebAssembly functions cannot return borrowed references or take dynamically sized vectors as function arguments. Try compiling the Rust to WebAssembly by adding the `#[wasm_bindgen]` attribute and take a look at the errors you get.
+
+```rust
+pub fn cells(u: &Universe) -> &[Cell] {
+    &u.cells
+}
+
+/// Pass a vector of indices to set only those cells to be alive.
+pub fn set_cells(u: &mut Universe, indices: Vec<u32>) -> &mut Universe {
+    let input_cells: Vec<Cell> = (0..u.width * u.height)
+        .map(|i| {
+            if indices.contains(&i) {
```

Comment:
Rather than iterating over all the universe's cells (and then iterating over each index for every cell, making the method `O(m*n)`), we can iterate over only the cells we are setting:

```rust
// Assuming we are taking row+col pairs as commented above...
for (row, col) in cells.iter().cloned() {
    let idx = get_index(row, col);
    self.cells[idx] = Cell::Alive;
}
```

If you want to hang on to some of the functional-ness you can also do a `map` over the input cells to get their indices before doing the mutation on `self.cells` :laughing: